articles/react-error-handling.md

```diff
@@ -0,0 +1,436 @@
+
+![product-search-error-boundary-reset-key.gif](/images/articles/react-error-handling/product-search-error-boundary-reset-key.gif)
+
+Did you get it? The first error is thrown as we set it then hitting the search button does not work but if we change the input's value and hit the button it will work. Well, we need to review the search button's handler, it is taking the inputRef's value and then setting it as productId value. If you take a look back at the outcome we are hitting the search button with the same productId (the state is not changing) so the ErrorBoundary does not have to reset its state due to resetKeys prop has not changed. On the other hand when we change the input's value it counts as a change and ErrorBoundary will reset it.
+
+So, the takeaway is to know that after a thrown error the resetKeys should be updated? Yeah, that's right. A comoon approach is to have a reset button (our ty again) inside the fallBack component so this button will reset the state and not re make the request as we do (I invite you to do the needed updates to run the common approach). But you know what? Let's keep having our current approach, we are warriors and we are not afraid of anything.
+
+![sweated-guy.jpg](/images/memes/sweated-guy.jpg)
+
+We have to make some changes if we want our app to be finally working. First of all let's remove the useEffect we can handle all directly on the getProductData function getting internally the input's value so this also implies removing its wrapper useCallback and the state productId, let's see how this looks like in code:
+
+```tsx
+function App() {
+  const [productData, setProductData] = React.useState<ProductData | null>(null)
+  const [error, setError] = useState<string | null>(null)
+  const [status, setStatus] = useState('idle')
+  const inputRef = React.useRef<HTMLInputElement>(null)
+
+  const handleSearch = () => {
+    getProductData()
+  }
+
+  const getProductData = () => {
+    const productId = inputRef.current?.value ?? ''
+
+    if (!productId) {
+      return
+    }
+
+    setStatus('pending')
+    makeRequest(productId)
+      .then((data) => {
+        setStatus('resolved')
+        setProductData(data)
+      })
+      .catch((error: Error) => {
+        setStatus('rejected')
+        setError(error.message)
+      })
+  }
+
+  const renderContent = () => {
+    //...
+  }
+
+  return (
+    <div>
+      <div>
+        <input type="text" ref={inputRef} />
+        <button onClick={handleSearch}>Search</button>
+      </div>
+      <div>
+        <ErrorBoundary 
+          FallbackComponent={ErrorFallback} 
+          onReset={getProductData}
+          resetKeys={[status]}
+        >
+          {renderContent()}
+        </ErrorBoundary>
+      </div>
+    </div>
+  )
+} 
+```
+
+![product-search-error-boundary-working.gif](/images/articles/react-error-handling/product-search-error-boundary-working.gif)
+
+We did it, all is working as expected (you can test the other scenarios by your own). 
```

Comment:
by !== on, in this case you should use on (using by usually implies there would be something after "your own")
```suggestion
We did it, all is working as expected (you can test the other scenarios on your own). 
```